packages/foundry/contracts/hooks/CommitMinerHook.sol

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "Balancer/BaseHooks.sol";
+import "Balancer/VaultGuard.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+
+contract CommitMinerHook is BaseHooks, VaultGuard, ReentrancyGuard {
+    struct Commit {
+        bytes32 hash;         // The commit hash
+        uint256 blockNumber;  // The block number when the commit was created
+        address pool;         // The pool that generated the commit
+        address swapper;      // Address of the user who performed the swap
+    }
+
+    IERC20 public immutable paymentToken;  // The ERC20 token used for both payments and rewards
+    address private immutable hookDeployer;  // Address of the hook deployer
+    Commit[] public commitBacklog;  // Array storing all commits
+    uint256 public feePercentage = 10;  // 10% fee to hook deployer, 30% swappers, 60% pool LPs
+
+    event CommitGenerated(address indexed pool, address indexed swapper, bytes32 commitHash, uint256 blockNumber);
+    event RandomnessRequested(address indexed requester, uint256 amountPaid, uint256 commitCount);
+    event FeeDistributed(address indexed pool, address indexed recipient, uint256 amount);
+
+    constructor(IVault vault, IERC20 _paymentToken) VaultGuard(vault) {
+        hookDeployer = msg.sender;
+        paymentToken = _paymentToken;  // The specified ERC20 token for all payments and rewards
+    }
+
+    function getHookFlags() public pure override returns (IHooks.HookFlags memory hookFlags) {
+        hookFlags.shouldCallBeforeSwap = true;
+    }
+
+    /// @inheritdoc IHooks
+    function onRegister(address factory, address pool, ...) public override onlyVault returns (bool) {
+        return true;
+    }
+
+    // Hook triggered before a swap to create a commit
+    function onBeforeSwap(
+        address pool,
+        uint256 amountIn,
+        uint256 amountOut,
+        address tokenIn,
+        address tokenOut,
+        address user
+    ) external override onlyVault {
+        bytes32 commitHash = keccak256(abi.encodePacked(block.timestamp, user, amountIn, amountOut, pool, gasleft()));
+        commitBacklog.push(Commit(commitHash, block.number, pool, user));
+
+        emit CommitGenerated(pool, user, commitHash, block.number);
+    }
+
+    // Randomness requester pays for a batch of commits using the specified paymentToken
+    function requestRandomness(uint256 commitCount, uint256 amountPaid) external nonReentrant {
+        require(commitCount <= commitBacklog.length, "Not enough commits available");
```

Comment:
This grows without bound, so there would have to be some kind of validation here in production. Like using a circular buffer instead of a simple array (1024 or 2048), and limiting the `commitCount`, or it would run out of gas computing it. The writes would also get cheaper once the buffer was full, vs. always being expensive.